### ISSUE TITLE
Filtering on map with subscript can crash or throw if the map is null

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10417]
This is a spinoff from #10399, which was about errors (and even crashes) when we have a filtering expression with a null subcript on a map column - e.g., `m[null] = 3` because null objects are being incorrectly deserialized. During that problem's debugging I found an more serious problem and am opening this issue for it:

Even if the expression is perfectly non-null, e.g., `m[2] = 3`, we can get similar deserialization errors (and theoretically, crashes) when the **data** in the table is null. For example, here is a reproducing test:

```python
def test_rabbit_hole_in_10399(cql, test_keyspace):
    schema = 'p text primary key, s set<int>, m map<int, int>'
    with new_test_table(cql, test_keyspace, schema) as table:
        cql.execute(f"INSERT INTO {table} (p,s) VALUES ('dog', {{1, 2}})")
        cql.execute(f"INSERT INTO {table} (p,s) VALUES ('cat', {{3, 4}})")
        assert list(cql.execute(f"SELECT p FROM {table} WHERE m[2] = 3 ALLOW FILTERING")) == []
```

You can see here we have data (two rows) where the `m` column wasn't set (i.e., null), and "m[2] = 3" should have simply not picked up those two rows. Instead, we get the following serious-looking marshaling error:

> E   cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.20.177.233:9042 datacenter1>: <Error from server: code=0000 [Server error] message="**marshaling error: read_simple_exactly - size mismatch (expected 4, got 1610612880)** Backtrace: 0x273a8be 0x273adb0 0x273b0b8 0x159be0f 0x159bb64 0x157d283 0x157d214 0x15633a8 0x15917a5 0x156342f 0x157bf28 0x1354b3b 0x1354889 0x1353cf3 0x13539ae 0x135111f 0x1347501 0x1724245 0x17247b8 0x1474418 0x1e05334 0x1e04c50 0x1e047c4 0x1e03b12 0x1e03913 0x1e037b9 0x1e03688 0x1e01017 0x1e00e82 0x2602ca4 0x2603f17 0x260330a 0x25b80e4 0x25b76e1 0xc3491e 0xc3390a /lib64/libc.so.6+0x27b74 0xc330ad

The backtrace is `cql3::selection::result_set_builder::restrictions_filter::do_filter()` calling `cql3::expr::is_satisfied_by()` calling `cql3::expr::(anonymous namespace)::equal()`. My unproven guess is that the code takes a null pointer and tries to deserialize it as an integer (expected size 4).

For some reason, the above reproducer no longer reproduces the bug if I remove the irrelevant "s" column. I don't know why that is. Maybe it's just a matter of luck (adding "s" just changes the data in memory a bit and converts a would-be-forgiven zero into bug-causing garbage).

CC @cvybhu @psarna @avikivity

**Comment by nyh** [comment]
I was wrong that we need the column "s" or the two rows to reproduce this bug. The reproduction is just random (probably relying on some random values in memory). The following even simpler test seems to fail in about half the runs, and worse - it also crashed Scylla on a segmentation fault just like #10399:

```python
def test_filtering_null_map_with_subscript(cql, test_keyspace):
    schema = 'p text primary key, m map<int, int>'
    with new_test_table(cql, test_keyspace, schema) as table:
        cql.execute(f"INSERT INTO {table} (p) VALUES ('dog')")
        assert list(cql.execute(f"SELECT p FROM {table} WHERE m[2] = 3 ALLOW FILTERING")) == []
```

**Comment by psarna** [comment]
Do we need to backport the fix somewhere?

**Comment by nyh** [comment]
As I answered Avi in https://github.com/scylladb/scylla/pull/10420, the specific evaluation function which I fixed in my patch was merged a month and a half ago, commit https://github.com/scylladb/scylla/commit/a8287a158ab06081725ed6087deb3c35f3898939 part of the merge https://github.com/scylladb/scylla/commit/9359a2caad3460ec1f23b70cdfdc3a83486e9ba0, so we surely had the same bug for a month and a half now so the fix as-is might be relevant to branch 5.0.

However, I don't know which other code we had earlier that implemented subscript maps - and whether it was susceptible to this or similar bugs. We can run the new tests from master against older branches to see how these null maps and subscripts behave there. In any case, sadly, if older versions will need to be fixed the fix will probably be different, because the expression-evaluation code fixed in these patches didn't exist in this form before a month and a half ago.

**Comment by nyh** [comment]
@avikivity @benipeled do you know how come this issue didn't get a "backport candidate" tag yet despite being closed 7 hours ago? I see other issues closed by the same patch, like #10399, did get a backport candidate tag! Maybe our bot has a bug where it only adds this tag to one issue fixed by a patch instead of all of them?
I'll add the tag manually now.

**Comment by benipeled** [comment]
AFAIK the backport-candidate script is looking for commits with the `Fixes #<num>` pattern,
I see you closed the commit that was linked to this issue, if you have no other merged commit with the required pattern the script will not gonna mark it as backport candidate.

**Comment by nyh** [comment]
@benipeled commit 808a93d29bc8e4335e111201f99e4b2296ccc8d0 does have "Fixes #10417". The merge commit 728479a6eae6b1c00b214b59bdf13f33288452cb should have also had it but had a typo. Is it possible you only look for Fixes in **merge** commits and not in individual commits?

**Comment by avikivity** [comment]
5.0 backport: 23a34d7e42fe25fd84cdc7ebc1270f761b6b793c
4.6 backport: 14e13ecbd476b35b176e0712ed8d656ab911b3fe
